core/factorization/ilu.cpp

```diff
@@ -0,0 +1,131 @@
+GKO_REGISTER_OPERATION(initialize_row_ptrs_l_u,
+                       par_ilu_factorization::initialize_row_ptrs_l_u);
+GKO_REGISTER_OPERATION(initialize_l_u, par_ilu_factorization::initialize_l_u);
+
+
+}  // namespace ilu_factorization
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<Composition<ValueType>> Ilu<ValueType, IndexType>::generate_l_u(
+    const std::shared_ptr<const LinOp> &system_matrix) const
+{
+    using CsrMatrix = matrix::Csr<ValueType, IndexType>;
+
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+
+    const auto exec = this->get_executor();
+    const auto host_exec = exec->get_master();
+
+    // Only copies the matrix if it is not on the same executor or was not in
+    // the right format. Throws an exception if it is not convertible.
+    auto local_system_matrix = CsrMatrix::create(exec);
+    as<ConvertibleTo<CsrMatrix>>(system_matrix.get())
+        ->convert_to(local_system_matrix.get());
+
+    // Add explicit diagonal zero elements if they are missing
+    exec->run(ilu_factorization::make_add_diagonal_elements(
+        local_system_matrix.get(), false));
+
+    // Compute LU factorization
+    exec->run(ilu_factorization::make_compute_ilu(local_system_matrix.get()));
+
+    // Separate L and U factors: nnz
+    const auto matrix_size = local_system_matrix->get_size();
+    const auto num_rows = matrix_size[0];
+    Array<IndexType> l_row_ptrs{exec, num_rows + 1};
+    Array<IndexType> u_row_ptrs{exec, num_rows + 1};
+    exec->run(ilu_factorization::make_initialize_row_ptrs_l_u(
+        local_system_matrix.get(), l_row_ptrs.get_data(),
+        u_row_ptrs.get_data()));
+
+    IndexType l_nnz_it{};
+    IndexType u_nnz_it{};
+    // Get nnz from device memory
+    host_exec->copy_from(exec.get(), 1, l_row_ptrs.get_data() + num_rows,
+                         &l_nnz_it);
+    host_exec->copy_from(exec.get(), 1, u_row_ptrs.get_data() + num_rows,
+                         &u_nnz_it);
+    auto l_nnz = static_cast<size_type>(l_nnz_it);
+    auto u_nnz = static_cast<size_type>(u_nnz_it);
+
+    // Init arrays
+    Array<IndexType> l_col_idxs{exec, l_nnz};
+    Array<ValueType> l_vals{exec, l_nnz};
+    std::shared_ptr<CsrMatrix> l_factor = matrix_type::create(
+        exec, matrix_size, std::move(l_vals), std::move(l_col_idxs),
+        std::move(l_row_ptrs), parameters_.l_strategy);
+    Array<IndexType> u_col_idxs{exec, u_nnz};
+    Array<ValueType> u_vals{exec, u_nnz};
+    std::shared_ptr<CsrMatrix> u_factor = matrix_type::create(
+        exec, matrix_size, std::move(u_vals), std::move(u_col_idxs),
+        std::move(u_row_ptrs), parameters_.u_strategy);
+
+    // Separate L and U: columns and values
+    exec->run(ilu_factorization::make_initialize_l_u(
+        local_system_matrix.get(), l_factor.get(), u_factor.get()));
+
+    return Composition<ValueType>::create(std::move(l_factor),
+                                          std::move(u_factor));
+}
```

Comment:
Here I'm a bit dubious in terms of benchmark quality. While most of that is required for Ginkgo specifics of the interface, if we are to benchmark the vendor libraries ilu factorization, then pretty much most of the relevant time would be comprised in `compute_ilu` no? It's the same in par_ilut with the `compute_l_u_factors` function being the important function AFAIK. All of these allocations and so on to wrap the data into CSR matrices that we then put in a composition are purely Ginkgo specifics. Therefore the `generate` time would have little relevance when benchmarking.

Reply:
The generate time would have little relevance, yes. But we can always just report the runtime of the `compute_ilu` operation (and I should note that the overall runtime is dominated by that and the triangular solve analysis phases)

Reply:
Yes, of course thanks to the loggers which report the time of separate operations. I think at the minimum we should have an immense warning somewhere, that the correct "generate" or factorization time at least for the factorizations should be taken from the specific `compute` kernels, and not the global generate time. That is different from say Jacobi, where the `generate` time is straight to the point a rather proper representation of the actual time.

Whereas here, due to the fact we want to separate the L U into Ginkgo CSR matrices among other quality checks, we can/could have a significant overhead which would blur the reality of the results.

Reply:
I will add some performance numbers before we merge this PR, as far as I remember, the separation into L and U was really negligible

Reply:
![overhead](https://user-images.githubusercontent.com/1693511/78304639-78e6c300-753f-11ea-9686-50571bbbf469.png)

The actual kernels (`factorization::*`) look to be around 5% of the total runtime, the allocations and copy we would need to do anyways. The only thing that kind of surprises me is the `free` overhead, but I am not sure how much we can do about that.

Reply:
Thanks for these plots that's an interesting picture. Copy seems OK but free and allocate look rather significant, in particular the free indeed. For ani7 which I guess is a bigger matrix, it looks like the proportion of these "extra" operations are not so insignificant anymore except for the V100. The three architectures really don't behave in the same way.

Reply:
Actually I accidentally added two different benchmark runs from the K20Xm to the plot, instead of correctly adding the Radeon7 timings. The plots should look like this:
![overhead-abs](https://user-images.githubusercontent.com/1693511/78309251-c3ba0800-754a-11ea-857d-b432d2498b92.png)
![overhead-rel](https://user-images.githubusercontent.com/1693511/78309254-c4529e80-754a-11ea-832d-0eebab05ab39.png)

Reply:
Wait - the generation time for ani7 on v100 is exceeding that of k20?

Reply:
Yes

Reply:
just that I understand it right - the timing includes the preparation of the triangular factors for the preconditioner application? In the performance graph you sent me, is it included there too?

Reply:
Yes, these are the complete preconditioner generation runtimes, containing the elimination of structural zeros (add_diagonal_elements), the ILU generation (compute_lu), the separation into lower and upper triangular factors (initialize*_l_u) and the solver analysis phases (generate). I was just as surprised about this as you are. The setup is
```
auto ilu_fact = std::shared_ptr<gko::LinOpFactory>(
    gko::factorization::Ilu<etype>::build().on(exec));
gko::preconditioner::Ilu<>::build()
     .with_factorization_factory(ilu_fact)
     .on(exec);
```

Reply:
hum... maybe we should exclude the preparation of the sparse systems in the plot for increasing ani dimension? How would it look like then? [https://www.overleaf.com/project/5e71c3074ed74a0001c2bbd8](https://www.overleaf.com/project/5e71c3074ed74a0001c2bbd8)

Reply:
As I was saying, I think for the sparselib case the `compute_lu` function is the most representative of the actual work. The problem is how relevant it is to compare against parilu's `compute_l_u_factors`, since this already writes in the separate factor instead of in place as sparselib do.

Reply:
Doesn't this actually just show that the hipSPARSE ILU generation is faster than the CUSPARSE one ? With also that your compute_l_u is also faster on the AMD gpus. Maybe, the applies are different ? As you say, the apply on the K20x is much slower than on the V100. Is it similar for the Radeon 7 ?

Reply:
The apply runtimes on the V100 and Radeon 7 are comparable, around 4ms. The astonishing thing is that apparently the analysis phases of ILU(0) and the triangular solves did not get any faster with at least 5 years of hardware development (The compute_lu kernel actually contains two phases: the analysis and the ILU solve. With the current setup, it is however not possible to separately measure their runtimes. Separating them into two kernels would be complicated, as this would require something similar to the SolveStruct)

Reply:
The compute_lu is a ParILU isnt it ? So by nature isnt that faster than the level scheduled ILU(0) analysis and trisolve phases ?

Reply:
No, this is still only the sparselib ILU, no ParILU in these plots

Reply:
Another interesting thing, for ani4 compute_lu (and everything else) seem to have shrinked quite a bit between V100 and K20. Weird that ani7 doesn't behave like that.